src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -111,12 +107,49 @@ static uint ComputeStringHash(string s)
         }
 
         /// <summary>Gets whether a given regular expression method is supported by the code generator.</summary>
-        private static bool SupportsCustomCodeGeneration(RegexMethod rm) =>
-            // The generator doesn't currently know how to emit code for NonBacktracking.
-            (rm.Options & RegexOptions.NonBacktracking) == 0;
+        private static bool SupportsCodeGeneration(RegexMethod rm)
+        {
+            RegexNode root = rm.Code.Tree.Root;
+
+            if (!root.SupportsCompilation())
+            {
+                return false;
+            }
+
+            if (ExceedsMaxDepthForSimpleCodeGeneration(root, allowedDepth: 40))
```

Comment:
I'm wondering if we should make the max depth configurable, although I'm not sure how that would be done for a source generator as something like an AppContext switch probably wouldn't work.

Reply:
I could add an environment variable.

Reply:
I'm not familiar enough with RegexNode yet to know how big of a default limit is 40, if we think it is very unlikely for folks to hit it then I'm fine with keeping it as is. If we expect some folks in advance scenarios to hit it, then it might be good to make it configurable, specially for cases where the pattern has a depth just over the limit like 41 or 42 since they would be missing out all of the perf gains just for a constrain protecting something that probably wouldn't make that much of a difference in their case.

Reply:
> I'm not familiar enough with RegexNode yet to know how big of a default limit is 40

Basically this would mean the regex has an expression nested in ~40 parentheses.

In the runtime-assets data set, this is what max depths look like (depth: number of expressions with that max depth):
```
         3: 4107
         4: 3592
         5: 2388
         6: 2012
         7: 1726
         8: 1270
         2: 1214
         9: 866
        10: 573
        11: 362
        12: 313
        13: 216
        14: 124
        15: 71
        16: 37
        18: 31
        17: 23
        19: 18
        23: 6
        22: 6
        20: 3
        21: 3
        25: 2
        26: 1
```

Reply:
I see, thanks for the context. By looking at that seems like reaching to 40 is going to not be very common, so I think it is fine to not make it configurable for now and we can revisit in the future if we ever get an issue logged for folks reaching this limit. Specially since there is no good way to make these configurations on the source generator (given env variable option would apply to all apps running on that machine). Only other alternative I can think of is a extra parameter to the generator attribute, but I don't think we should add that complexity now.